diff --git a/net/core/dev.c b/net/core/dev.c
index a4bf943..9247698 100644
--- a/net/core/dev.c
+++ b/net/core/dev.c
@@ -4899,7 +4899,7 @@ static int dev_ifsioc_locked(struct net *net, struct ifreq *ifr, unsigned int cm
 		 * is never reached
 		 */
 		WARN_ON(1);
-		err = -ENOTTY;
+		err = -EINVAL;
 		break;
 
 	}
@@ -5173,7 +5173,7 @@ int dev_ioctl(struct net *net, unsigned int cmd, void __user *arg)
 		/* Set the per device memory buffer space.
 		 * Not applicable in our case */
 	case SIOCSIFLINK:
-		return -ENOTTY;
+		return -EINVAL;
 
 	/*
 	 *	Unknown or private ioctl.
@@ -5194,7 +5194,7 @@ int dev_ioctl(struct net *net, unsigned int cmd, void __user *arg)
 		/* Take care of Wireless Extensions */
 		if (cmd >= SIOCIWFIRST && cmd <= SIOCIWLAST)
 			return wext_handle_ioctl(net, &ifr, cmd, arg);
-		return -ENOTTY;
+		return -EINVAL;
 	}
 }
 
