diff --git a/base/message_pump_libevent.cc b/base/message_pump_libevent.cc
index ff0d40d..8c5fc18 100644
--- a/base/message_pump_libevent.cc
+++ b/base/message_pump_libevent.cc
@@ -6,6 +6,7 @@
 
 #include <errno.h>
 #include <fcntl.h>
+#include <unistd.h>
 
 #include "base/auto_reset.h"
 #include "base/compiler_specific.h"
diff --git a/chrome/browser/diagnostics/diagnostics_main.cc b/chrome/browser/diagnostics/diagnostics_main.cc
index edc1c55..2ee342e 100644
--- a/chrome/browser/diagnostics/diagnostics_main.cc
+++ b/chrome/browser/diagnostics/diagnostics_main.cc
@@ -6,8 +6,8 @@
 
 #if defined(OS_POSIX)
 #include <stdio.h>
-#include <unistd.h>
 #endif
+#include <unistd.h>
 
 #include <string>
 
diff --git a/chrome/browser/policy/policy_path_parser_linux.cc b/chrome/browser/policy/policy_path_parser_linux.cc
index 2f9ea26..cddd58b 100644
--- a/chrome/browser/policy/policy_path_parser_linux.cc
+++ b/chrome/browser/policy/policy_path_parser_linux.cc
@@ -3,6 +3,7 @@
 // found in the LICENSE file.
 
 #include <pwd.h>
+#include <unistd.h>
 
 #include "chrome/browser/policy/policy_path_parser.h"
 
diff --git a/ipc/ipc_channel.h b/ipc/ipc_channel.h
index bfb4da7..16647d0 100644
--- a/ipc/ipc_channel.h
+++ b/ipc/ipc_channel.h
@@ -7,6 +7,7 @@
 #pragma once
 
 #include <string>
+#include <sys/types.h>
 
 #include "base/compiler_specific.h"
 #include "ipc/ipc_channel_handle.h"
diff --git a/ipc/ipc_platform_file.cc b/ipc/ipc_platform_file.cc
index b5ec7be..f704bd2 100644
--- a/ipc/ipc_platform_file.cc
+++ b/ipc/ipc_platform_file.cc
@@ -4,7 +4,7 @@
 
 #include "ipc/ipc_platform_file.h"
 
-#if defined(OS_ANDROID)
+#if defined(OS_ANDROID) || defined(OS_POSIX)
 #include <unistd.h>
 #endif
 
diff --git a/seccompsandbox/library.h b/seccompsandbox/library.h
index 5f0d781..18eb531 100644
--- a/seccompsandbox/library.h
+++ b/seccompsandbox/library.h
@@ -12,6 +12,7 @@
 #include <string>
 #include <string.h>
 #include <sys/mman.h>
+#include <stdlib.h>
 
 #include "maps.h"
 
diff --git a/ui/base/l10n/l10n_util.cc b/ui/base/l10n/l10n_util.cc
index 28dbc43..8b1998e 100644
--- a/ui/base/l10n/l10n_util.cc
+++ b/ui/base/l10n/l10n_util.cc
@@ -5,7 +5,7 @@
 #include "ui/base/l10n/l10n_util.h"
 
 #if defined(USE_X11)
-#include <glib/gutils.h>
+#include <glib-2.0/glib.h>
 #endif
 
 #include <algorithm>
